fs/src/io_uring/memory.rs

```diff
@@ -14,64 +15,67 @@ use {
 // and chunk it in slices of up to 1G each.
 const FIXED_BUFFER_LEN: usize = 1024 * 1024 * 1024;
 
-pub enum LargeBuffer {
-    Vec(Vec<u8>),
-    HugeTable(PageAlignedMemory),
-}
-
-impl Deref for LargeBuffer {
-    type Target = [u8];
+#[derive(Debug)]
+pub struct AllocError;
 
-    fn deref(&self) -> &Self::Target {
-        match self {
-            Self::Vec(buf) => buf.as_slice(),
-            Self::HugeTable(mem) => mem.deref(),
-        }
+impl From<AllocError> for io::Error {
+    fn from(_: AllocError) -> io::Error {
+        io::Error::other("failed to allocate page aligned memory")
     }
 }
 
-impl DerefMut for LargeBuffer {
-    fn deref_mut(&mut self) -> &mut Self::Target {
-        match self {
-            Self::Vec(buf) => buf.as_mut_slice(),
-            Self::HugeTable(mem) => mem.deref_mut(),
-        }
-    }
+enum AllocationMethod {
+    Std(Layout),
+    Mmap,
 }
 
-impl AsMut<[u8]> for LargeBuffer {
-    fn as_mut(&mut self) -> &mut [u8] {
-        match self {
-            Self::Vec(vec) => vec.as_mut_slice(),
-            LargeBuffer::HugeTable(mem) => mem,
-        }
-    }
+/// Memory buffer aligned to the system page size.
+///
+/// `PageAlignedMemory.ptr` is always aligned to the system page size.
+/// `PageAlignedMemory.len` may or may not be aligned to the page size.
+/// When it is allocated, if the `len` is smaller than the page size,
+/// only `ptr` is aligned to the page size.
+/// If a huge page was used to allocate, then
+/// both `ptr` and `len` are aligned to the page size.
```

Comment:
maybe something like

```suggestion
/// Memory buffer whose base address is aligned to the system page size.
///
/// `ptr` is always aligned to the system page size.
/// `len` is not guaranteed to be page-aligned.
///
/// On allocation:
/// - If `len` is smaller than the system page size, only `ptr` is
///   page-aligned.
/// - Otherwise allocation might use huge pages and if it succeeds, both `ptr` and `len` are
///   aligned to the system page size.
```

Reply:
a3d3526576943fcd44df648f24b325395b9b9a7c